src/plumpy/processes.py

```diff
@@ -1307,10 +1308,35 @@ def execute(self) -> Optional[Dict[str, Any]]:
 
         :return: None if not terminated, otherwise `self.outputs`
         """
+        from plumpy.reentry import get_runner
+
         if not self.has_terminated():
-            self.loop.run_until_complete(self.step_until_terminated())
+            coro = self.step_until_terminated()
+            # FIXME: define loop in process init is problematic, since it may get the main thread
+            # event loop and close it unexpectly. I think should be only two type of entries where
+            # the event loop definition is clear.
+            # A. if there already a running loop, create a thread with event loop to offload run, and support nested
+            # B. if there is no loop, create its own that can nested using greenlet
+            loop = self.loop or asyncio.get_running_loop()
```

Comment:
For aiida-core use case, I may want to distinguish the use of thread by checking if it is launch using `run` or using `submit`. If it `submit`, it means it is going to push to the same event loop event it can block. If it is `run` then it always spawn a thread to run.